src/scales/scale.time.js

```diff
@@ -51,246 +175,254 @@ module.exports = function(Chart) {
-							}
+					// Let's consider that all data have the same format.
+					if (helpers.isObject(data[0])) {
+						timestamps[i] = [];
 
-							timestamp = dataMoment.valueOf();
-							dataMin = Math.min(timestamp, dataMin);
-							dataMax = Math.max(timestamp, dataMax);
-							timestamps[dataIndex] = timestamp;
+						for (j = 0, jlen = data.length; j < jlen; ++j) {
+							timestamp = parse(data[j], me);
+							min = Math.min(min, timestamp);
+							max = Math.max(max, timestamp);
+							timestamps[i][j] = timestamp;
 						}
-					});
+					} else {
+						timestamps[i] = labels.slice(0);
+					}
 				} else {
-					// We have no x coordinates, so use the ones from the labels
-					timestamps = parsedData.labels.slice();
+					timestamps[i] = [];
 				}
+			}
 
-				parsedData.datasets[datasetIndex] = timestamps;
-			});
-
-			me.dataMin = dataMin;
-			me.dataMax = dataMax;
-			me._parsedData = parsedData;
+			// Enforce limits with user min/max options
+			min = parse(options.time.min, me) || min;
+			max = parse(options.time.max, me) || max;
+
+			// In case there is no valid min/max, let's use today limits
+			min = min === MAX_INTEGER ? +moment().startOf('day') : min;
+			max = max === MIN_INTEGER ? +moment().endOf('day') + 1 : max;
+
+			me._model = {
+				datasets: timestamps,
+				horizontal: me.isHorizontal(),
+				labels: labels.sort(sorter),    // Sort labels **after** data have been converted
+				min: Math.min(min, max),        // Make sure that max is **strictly** higher ...
+				max: Math.max(min + 1, max),    // ... than min (required by the lookup table)
+				offset: null,
+				size: null,
+				table: []
+			};
 		},
+
 		buildTicks: function() {
 			var me = this;
+			var model = me._model;
+			var min = model.min;
+			var max = model.max;
 			var timeOpts = me.options.time;
-
-			var minTimestamp;
-			var maxTimestamp;
-			var dataMin = me.dataMin;
-			var dataMax = me.dataMax;
-
-			if (timeOpts.min) {
-				var minMoment = timeHelpers.parseTime(me, timeOpts.min);
-				if (timeOpts.round) {
-					minMoment.startOf(timeOpts.round);
+			var ticksOpts = me.options.ticks;
+			var formats = timeOpts.displayFormats;
+			var capacity = me.getLabelCapacity(min);
+			var unit = timeOpts.unit || timeHelpers.determineUnit(timeOpts.minUnit, min, max, capacity);
+			var majorUnit = timeHelpers.determineMajorUnit(unit);
+			var ticks = [];
+			var i, ilen, timestamp, stepSize;
+
+			if (ticksOpts.source === 'labels') {
+				for (i = 0, ilen = model.labels.length; i < ilen; ++i) {
+					timestamp = model.labels[i];
+					if (timestamp >= min && timestamp <= max) {
+						ticks.push(timestamp);
+					}
 				}
-				minTimestamp = minMoment.valueOf();
-			}
-
-			if (timeOpts.max) {
-				maxTimestamp = timeHelpers.parseTime(me, timeOpts.max).valueOf();
+			} else {
+				stepSize = helpers.valueOrDefault(timeOpts.stepSize, timeOpts.unitStepSize);
+				stepSize = stepSize || timeHelpers.determineStepSize(min, max, unit, capacity);
```

Comment:
You could probably make these a single statement:

	stepSize = helpers.valueOrDefault(timeOpts.stepSize, timeOpts.unitStepSize)
		|| timeHelpers.determineStepSize(min, max, unit, capacity);